### ISSUE TITLE
Invalid response received by Tweakwise server, response code is not 200.

### ISSUE THREAD

**Issue by jeroen-severein** [EmicoEcommerce/Magento2Tweakwise-archived#138]
For two of our customers we see this error tens of thousands of times in our logs.
Invalid response received by Tweakwise server, response code is not 200. Request "https://gateway.tweakwisenavigator.com/recommendations/product/72cbc89a/2/100033626"
The invalid response that is given is: 404 Product not found. The last ID in the URL is different per error, but they are always product id's that don't exist in Tweakwise.

Environment
--
* PHP Version: 7.2
* Magento Version: 2.3.2
* Tweakwise Version: 2.3.4
* Tweakwise Export Version: 1.6.3
* Magento Deploy Mode: production

**Comment by edwinljacobs** [comment]
Hi,

Judging by the url and the fact that you mentioned that they are not available in tweakwise, product with magento entity id 3626 is not present in the export feed (this is generated by the export module) for store id 3. Why that is is not clear, apparently the recommendation rule does trigger for that product. This leads me to believe that there are links to this product that do not come from navigation or search (since this product is not available in the feed for that particular store). The product is apparently enabled and also visible on the frontend and for some reason not in the feed. There is an update available for the export module v2.1.0 (note that this is a major change and could break any custom extensions you might have) you could try to check if that fixes the issue but my guess would be that it is something else.

Is there a "non navigation" way to reach those products? Maybe they are not in the feed for a specific reason?

With kind regards

**Comment by jeroen-severein** [comment]
Hi,

Ok that makes sense. I think the problem is then that we use Grouped Products as the Product Detail Page in Magento 2. So for example a T-shirt, where the child products are the different sizes. In case of this URL, this is the page for product 3626 and store id 3: https://www.stanno.com/nl/481366-8300-volare-match/ It does show a Tweakwise recommendations block on this page. But only the child/simple products are exported to Tweakwise.

Is there a way to export these grouped products? Or perhaps it would be better that the recommendations block is loaded by the child product id's and not the grouped product?

**Comment by edwinljacobs** [comment]
Hi,

You mention that only the childs are available in the feed, any idea why that is? It could be that this is a bug in the export module (not sure though), the grouped products should be in the feed to be honest. There where some issues with determining stock status of certain product types in export versions < 2. I know this is relatively hard to debug since the export module does not use magento's repositories for getting product data. You could try the following on your development machine

1) In vendor/emico/tweakwise-export/Model/Write/Products.php, change method ```exportStore``` to
```
protected function exportStore(Writer $writer, XmlWriter $xml, Store $store): void
    {
        if ($store->getId() != 3) {
            return;
        }
        $this->iterator->setStore($store);
        // Purge iterator entity ids for each store
        $this->iterator->setEntityIds([3626]);

        foreach ($this->iterator as $index => $data) {
            $this->writeProduct($xml, $store->getId(), $data);
            // Flush every so often
            if ($index % 100 === 0) {
                $writer->flush();
            }
        }

        // Flush any remaining products
        $writer->flush();
    }
```
So you have only that product to worry about (note that this example is based on export version > 2 but idea is clear). You could do a similar thing to vendor/emico/tweakwise-export/Model/Write/Categories.php (just early return in the exportStore method so that you don't have to wait for all the categories to be processed).

Depending on the installed export version in class vendor/emico/tweakwise-export/Model/Write/Products/ExportEntity.php
there are methods ```shouldExport``` and ```shouldProcess``` (shouldProcess is only available when emico/tweakwise-export "^2" is installed. If these methods return false at some point then the product will not be available in the feed.

In export version < 2 the method shouldExport gets called twice once before all product data is collected and once after all the data is collected. In version 2.0 and onwards method shouldProcess is called once before all product data is collected and after that is done method shouldExport is called to check if the product should be in the feed.

All the classes in vendor/emico/tweakwise-export/Model/Write/Products/CollectionDecorator are responsible for adding data to the export entities. After all the data is collected method shouldExport gives the final verdict (in the feed or not in the feed).

I strongly suggest updating to 2.X if that does not give any problems, in version 2.X there exists a class for dealing solely with grouped products (vendor/emico/tweakwise-export/Model/Write/Products/ExportEntityGrouped.php) This makes it more clear. 

Hope this helps, if you have any questions feel free to ask

**Comment by jeroen-severein** [comment]
Hi,

Thanks for the debugging info. I am now a bit further in what is going on. It seems I had found the wrong product in my previous comment. It's actually this one: https://www.stanno.com/nl/481366-1400-volare-match/

So it seems the reason that these products are missing is because they are grouped products that don't have any children with stock. I assume this is because we have this set to No:
![Screen Shot 2021-01-04 at 11 02 51](https://user-images.githubusercontent.com/50044827/103523628-71609b80-4e7c-11eb-99b5-bbf9ca618cf3.png)

My question is if we set this to Yes what will happen? I assume these missing products will be added to the feed. But what behaviour will this cause? Ideally it will show the upsell/cross-sell blocks on the out of stock product page. Because I think there is value in showing the customer these blocks, so that they have alternatives when a product is out of stock. 

However we don't want these out of stock products to show up in the search feed or in the crossell/upsell blocks for other products. Is this something the Magento2 module does correct automatically or do we need to add extra configuration to the search feed and crossell/upsell profiles in Tweakwise?

**Comment by edwinljacobs** [comment]
Hi,

That setting controls if data from child products should be exported on the parent product. It does not decide if the parent product ends up in the feed. For some extra information on the feed take a look at the readme from https://github.com/EmicoEcommerce/Magento2TweakwiseExport here you can find some extra explanation about the feed.

You mentioned that this product does not have any in stock children. Maybe something goes wrong in 
```
/**
     * @return bool
     */
    protected function shouldExportByStock(): bool
    {
        if ($this->stockConfiguration->isShowOutOfStock($this->store->getId())) {
            return true;
        }

        return $this->getStockItem() ? (bool) $this->getStockItem()->getIsInStock() : false;
    }
```
This method can be found in class ExportEntity (vendor/emico/tweakwise-export/Model/Write/Products/ExportEntity.php)
The first if determines if your catalog configuration allows visible out of stock products.
```
 if ($this->stockConfiguration->isShowOutOfStock($this->store->getId())) {
            return true;
        }
```
Then a simplified stock item is retrieved and checked if it is in stock, this simplified stock item is constructed during the export.
The reason this is a different stock item is because magento released its new inventory management in 2.3, we also need to be compatible with 2.2 (sadly). You could check if that simplified stock item is available.

**Comment by jeroen-severein** [comment]
Hi,

Ok great, that makes sense. We have the isShowOutOfStock set to No:
![Screen Shot 2021-01-05 at 11 59 40](https://user-images.githubusercontent.com/50044827/103638780-abe63900-4f4d-11eb-91c2-3e963905f9ed.png)
And the stockItem is returned and the combined isInStock value is correctly determined as 0, so that is why they aren't being exported. If I change this setting to Yes it does get exported. And we would only need to make a change in the front-end view of the product detail page to keep the same layout.

So back to one of my previous questions:
What behaviour will this cause on the side of Tweakwise? Ideally it will show the upsell/cross-sell blocks on the out of stock product page. Because I think there is value in showing the customer these blocks, so that they have alternatives when a product is out of stock. However we don't want these out of stock products to show up in the search feed or in the crossell/upsell blocks for other products. Is this something Tweakwise/Magento2 module does correct automatically or do we need to add extra configuration to the search feed and crossell/upsell profiles in Tweakwise?

**Comment by edwinljacobs** [comment]
Hi,

The answer to your question depends on which setting you mean. 

In the case "Export out of stock combined children" this setting does not add or exclude products from the export feed. Given a composite product, say a configurable t-shirt with colors "blue" and "red" where "red" is out of stock. If you set this setting to "Yes" then the t-shirt is exported with color values blue and red, which means that if the user filters on color red the t-shirt will show up. If you change the setting to "No" the t-shirt will be exported with color value "Blue" meaning that if the user filters on "Red" the t-shirt will not show up. This setting does not determine which products end up in the feed, it merely determines for each composite product which data is added to that particular product in the export.

In the case of "Display out of stock products" it does what you expect it to do. If this is set to "yes" then Magento native will show you out of stock products in search results and in layered navigation. If this is set to "no" then it wont. We need to take this setting into account, tweakwise can't show you any products it doesn't know about and as such if this setting is set to yes we need to export the out of stock products. However if this setting is set to "No" we don't export these products otherwise they would end up in the search results and in layered navigation. So this setting determines to an extent which products show up in the feed.

I think the problem here is that 
1) In your case you do not want out of stock products in navigation and search, as such, "Display out of stock products" is set to no, the product does not end up in the feed. This is correct based on your last comment
2) For one reason or another a user or some crawler (this seems more likely since the product cannot be reached from navigation or search) ends up on a out of stock product, the recommendation rule triggers and requests recommendations which tweakwise cannot find because the product is not in the feed, in other words tweakwise knows nothing of this product. En error is thrown and ends up in the log files. It should be the case that the magento standard recommendations are still shown, you can check that at: https://github.com/EmicoEcommerce/Magento2Tweakwise/blob/bb3897cb73deead7ff4f698b8864b216f1a33f89/Block/Catalog/Product/ProductList/Upsell/Plugin.php#L42
Here proceed() is called in case the api fails to give a valid response. To answer your question, the module does not auto correct this.

So here is my view on the matter, recommendations are triggered when a product page is reached, a priori it is unclear if this product is known to tweakwise (as explained above). If it is not known in tweakwise this results in an ApiException which is logged and then handled by magento, it should render magento native upsells or related products. So in some sense this exception is "expected". 

There are two options in my opinion.

1) We could reduce the log level of this particular exception and the log would not cluttered with these messages. The magento standard relateds and upsells could be used to show alternatives for that product. I currently don't see a way to give the user tweakwise recommendations in this case. We would have to use some other "similar" product to get the recommendations in this case and there is currently no way to do that.

2) The other option is that you do export the out of stock products (set "Display out of stock products" to "yes") this will solve the issue but it will result in out of stock products in search results and in navigation. This is something you don't want however there is a way to circumvent this issue. It is possible to add hidden filters to a navigator request and tweakwise will only return results which matches this hidden filter. You could use this to filter out products which are not in stock. This could be an option. To see this in action you could take a look at https://github.com/EmicoEcommerce/Magento2Tweakwise/blob/bb3897cb73deead7ff4f698b8864b216f1a33f89/Model/Catalog/Layer/NavigationContext.php#L215
Here a filter is added to match the correct visibility settings for this request (when it is a search request we only want products that have magento visibility "Search" or "Catalog,Search"), note that the visibility filter should be added to the template in the navigator for this to work. As a side note, I'm currently working on something which makes it easier to add these hidden filters. 

The only downside to this last approach is that you would have to update all your tweakwise templates with a hidden filter "out of stock filter" and write some custom code to add this filter to all SearchRequests and ProductNavigationRequests.
 
I hope this gives some explanation and answers your questions.
I think there is a case for lowering the log level of these Exceptions however there is also a case for not lowering it because if we do it is unclear that this happens. I am not yet sure on how to approach this.

With kind regards

**Comment by jeroen-severein** [comment]
Hi,

Yes I'm also not sure on what would be the best solution for us. Updating the templates seems complex if there is no way of filtering these products from being returned through Tweakwise filters/algorithms. I feel that having recommendations for these kind of products would be a nice feature to have, so perhaps that something for us to take a look at in the future. The native Magento $proceed probably does trigger but for us it's empty, because those attributes aren't filled.

I suppose we could also just not load the crossell and upsell blocks when a product is out of stock. Then visually it would be the same for the customer, but at least the exception wouldn't be triggered. However I do see it occurring for another customer where we do send the out of stock products. It seems in that case the products are added faster than we export them to Tweakwise. So it only occurs for one day and then in the nightly export the missing products are added and it no longer occurs for those products. But for this customer products are added on a daily basis. I guess we could generate the feed more often there, provided this doesn't cause performance problems.

But to some extend I think it does make sense to change the log level for this. Because if a product is not exported yet or it's out of stock do seem like normal events that in our case don't require action from us, because either the block is empty or the fallback is shown or the product is exported later.

The main problem we have with it currently is that it's logged as an error with the URL in the message. Because of this our error monitoring tool Bugsnag sees each different url as a different problem. And because there are 17 shops x multiple recommendations profiles x thousands of products this triggers a lot of alerting for new occurrences. To the extend that it prevents us from manually marking these as ignored and it prevents us from noticing real problems faster.

So for us it would help a lot if for example 404 Product not found was logged as a warning or notice and that the URL is logged through the $context array, instead of in the message. That way for now our alerting wouldn't trigger and we would capture it as one ticket that we can ignore.

But I don't know if this is something that makes sense for all users and is something that you will implement? Or that this is so specific for us that we should create overrides for this.

**Comment by edwinljacobs** [comment]
I took the liberty of creating the override for you, I am not yet sure if this (or something like it) will end up in the module I'll have to think about it. Feel free to modify it to your needs. An obvious disclaimer, do test this if you want to use it. Its a pragmatic solution, nothing fancy
[Tweakwise.zip](https://github.com/EmicoEcommerce/Magento2Tweakwise/files/5770985/Tweakwise.zip)

I chose log level debug, again feel free to modify and change it to your needs (or not use it at all if it does not meet your standards). I hope it works in your case!

With kind regards
